### ISSUE TITLE
Adding ECDH to SSH/Conch

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8325]
|[<img alt="the0id's avatar" src="https://avatars.githubusercontent.com/u/19194467?s=50" width="50" height="50">](https://github.com/the0id)| @the0id reported|
|-|-|
|Trac ID|trac#8325|
|Type|enhancement|
|Created|2016-05-10 16:45:59Z|

UPDATE: I've created a new ticket [#8370](https://github.com/twisted/twisted/issues/8370) that includes ECDSA* and should supersede this ticket.

Here is my proposal to add ECDH to Conch.  It does not include ECDSA support (yet) so RSA/DSA keys are still required.

I ran the trials and it does fail on two tests.
The first one fails in twisted.scripts.test.test_tap2deb.TestTap2DEB.test_basicOperation

on:
"twisted.cred.strcred.InvalidAuthType: sshkey

twisted.cred.test.test_strcred.SSHCheckerTests.test_isChecker"

This test fails on an clean pull of Twisted so I assume this is not (directly) related to my code.

The second failure stems from apparent miss-matched host keys.  I looked into this and was unable to determine the cause.  I saw in the test that it appears to be trying to test connections to "ssh.example.com" which could explain the problem, or it could be that my development system doesn't have the host keys set up correctly.

I have done extensive testing with both server and client sides of Conch with my patch and have had no problems with connecting or any keys. So I'm submitting the patch even though it fails this test in hopes either it's an unrelated error on my part, or someone can help me identify the problem.

Attachments:

* [factory.py.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/eaa/eaa61686aca9e4913ee50946e464a3abb49d164c/232d1109932947a0ad54cdb60ba1704ce574e431.patch) (598 bytes) - added by the0id on 2016-05-10 16:47:27Z - The patch to conch/factory.py
* [_kex.py.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/eaa/eaa61686aca9e4913ee50946e464a3abb49d164c/5864b6ae3be40a647b4a69be6e0aa1de7cb09adf.patch) (4117 bytes) - added by the0id on 2016-05-10 16:47:48Z - The patch to _kex.py
* [transport.py.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/eaa/eaa61686aca9e4913ee50946e464a3abb49d164c/4f6281a84a17bd198b0ff65d6fc77532e40023af.patch) (12567 bytes) - added by the0id on 2016-05-10 16:48:06Z - The patch to transport.py
* [unittests](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/eaa/eaa61686aca9e4913ee50946e464a3abb49d164c/7789c430c9b593bde2fd07f795a8d54c7416d431) (1230250 bytes) - added by the0id on 2016-05-10 16:50:40Z - The output of trial twisted unit tests
* [bundle.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/eaa/eaa61686aca9e4913ee50946e464a3abb49d164c/f8a0e18d47bc1a5c63a80cb03fbbb7b4803f434d.patch) (17174 bytes) - added by the0id on 2016-05-11 13:33:58Z - Patches for transport.py, _kex.py, and factory.py together.

<details><summary>Searchable metadata</summary>

```
trac-id__8325 8325
type__enhancement enhancement
reporter__the0id the0id
priority__normal normal
milestone__None None
branch__ 
branch_author__None None
status__closed closed
resolution__fixed fixed
component__conch conch
keywords__None None
time__1462898759038604 1462898759038604
changetime__1503074974445324 1503074974445324
version__None None
owner__the0id the0id
cc__z3p
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_2"></a>|@hawkowl commented|
|-|-|

(not a review, just adding a note before bed :) )

Hi, the tap2deb failure is unrelated (and tap2deb is getting removed, see [#8326](https://github.com/twisted/twisted/issues/8326)), so don't worry about that :D

Also, once you've finished uploading your patches, add the "review" keyword and it'll get added to our review queue (https://twistedmatrix.com/trac/report/25 -- central place so the reviewers know where to look) :)

Thanks for your patch and your interest in improving Twisted!

**Comment by twisted-trac** [comment]
|[<img alt="the0id's avatar" src="https://avatars.githubusercontent.com/u/19194467?s=50" width="50" height="50">](https://github.com/the0id)<a name="note_4"></a>|@the0id commented|
|-|-|

No problem, though save your thanks until after the review. :)

**Comment by twisted-trac** [comment]
|[<img alt="the0id's avatar" src="https://avatars.githubusercontent.com/u/19194467?s=50" width="50" height="50">](https://github.com/the0id)<a name="note_5"></a>|@the0id commented|
|-|-|

And I found a bug in what I submitted.  It looks like I didn't roll back my ecdsa experimentation as much as I thought.

Line 491 should NOT be changed with the patch. It should remain:
supportedPublicKeys = [b'ssh-rsa', b'ssh-dss']

If you remove the file I can upload a fixed version.

Thanks.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_6"></a>|@hawkowl set owner to @the0id|
|-|-|

Hi, 

You can upload a patch with a -2 and it should work. Also, if you can submit all the changes as one diff (to make it easier to apply), that'd be great!

**Comment by twisted-trac** [comment]
|[<img alt="the0id's avatar" src="https://avatars.githubusercontent.com/u/19194467?s=50" width="50" height="50">](https://github.com/the0id)<a name="note_7"></a>|@the0id commented|
|-|-|

No problem, added as bundle.patch.  It also includes the above fix.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_9"></a>|@hawkowl commented|
|-|-|

This should be up in review, so people see it :)

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_10"></a>|@rodrigc set status to `closed`|
|-|-|

Moving this over to [ticket:8370]

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_11"></a>|@adiroiban commented|
|-|-|

I think that merging this as 2 separate jobs is preferred. Smaller patches are easier to review.

the0id if you are still interested to push this code in Twisted, I would be happy to review it.

**Comment by twisted-trac** [comment]
|[<img alt="the0id's avatar" src="https://avatars.githubusercontent.com/u/19194467?s=50" width="50" height="50">](https://github.com/the0id)<a name="note_12"></a>|@the0id commented|
|-|-|

Thanks, it appears that glyph agrees with you and suggested I do the same on the other ticket.

Since there's already a pull request for the other ticket, I went ahead and rolled back the ECDSA portions of the code and pushed up the changes.

https://github.com/twisted/twisted/pull/387

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_13"></a>|@adiroiban commented|
|-|-|

Thanks for your contribution.

Please note that all changes needs to come with 100% tests... both unit tests and OpenSSH integration tests.

In this way we make sure the changes work across all the supported platforms and they will also help prevent regression in the future.

Also, creating a dedicated PR will not hurt, as we will have a clean start :)

When writing new code you should add the minimum public API. Any code which should not be used outside of the twisted project should have names starting with underscored.

Let me know if you need help with the tests and API design.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_14"></a>|@adiroiban set status to `reopened`|
|-|-|

Hm... reopening this

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)|@adiroiban removed owner<br>@adiroiban set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_16"></a>|@adiroiban set owner to @the0id|
|-|-|

I think that we should wait for the EC patch to be merged and then we can take a look at this branch.

In the same time, the previous PR is not longer targeting this ticket.

So, nothing to review for now.

Looking forward for having this functionality in Twisted :)

Many thanks for your help!

Once there is a PR for ECDH support, please put it back for review.

**Comment by twisted-trac** [comment]
|[<img alt="the0id's avatar" src="https://avatars.githubusercontent.com/u/19194467?s=50" width="50" height="50">](https://github.com/the0id)<a name="note_17"></a>|@the0id commented|
|-|-|

No problem, I'm glad I was able to contribute. In hindsight I should have labeled my branches under this ticket number, but oh well. :)

The latest pull request is https://github.com/twisted/twisted/pull/432 and it's now passing the build tests.  As I was pushing in the final changes it looks like the main branch got updated so my branch is a little out of date.  I tried merging it on Github but for some reason was was doing incremental merges, and each one spawned a new test.

Not all of the tests have completed yet, but the only difference in the tests now are the upstream merges, which didn't cause any conflicts locally on my machine.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_18"></a>|@adiroiban commented|
|-|-|

Since this is still in the review queue, here is my review :)

-------

1. I don't understand the changes from twisted/conch/ssh/keys.py.

I think that we should have a separate ticket in which we add support for EC to twisted.conch.ssh.keys... so that we can generate/read/parse them

and this can be shared with ECDSA part

and we can look at exposing the ECDSA key in scripts/ckeygen


2. Also to reduce the size of this ticket, maybe we can have 'hmac-sha2-384' in a separate ticket... as it should be merged much faster and is not related to EC


3. For the duplicate code 

```
shortKex = re.search(b"(nist[kpbt]\d{3})$", self.kexAlg).group(1)
curve = keys.curveTable[shortKex]
ecPriv = ec.generate_private_key(curve, default_backend())
ecPub = ecPriv.public_key()
encPub = ecPub.public_numbers().encode_point()
theirECPub = ec.EllipticCurvePublicNumbers.from_encoded_point(curve, pktPub).public_key(default_backend())
```

I was hopting that we can just use twisted.conch.ssh.keys.Key.fromString ... or some help so that we can remove the code duplication in _ssh_KEX_ECDH_INIT and  ssh_KEXINIT

Also, in this way we might be able to have `curveTable` private.


4. What is the point of the oidTable ?


5. In ssh_KEXINIT do we need self.curve , self.ecPriv , self.ecPub as public members?


6. The change from twisted/conch/ssh/factory.py is missing a test :)

7. Also, instead of a method like `isFixedGroup` maybe we can have 'isVariableGroup` ... so if later we add Ed25519 or other method we don't have to create a new method for them. 

8. We also need OpenSSH interoperatiblity tests

9. I don't see a test for `b'bad host key'` or UnsupportedAlgorithm

--------

we might not be able to move all the code inside the Key object, but I think that the idea is to pipe all the keys received from SSH into the Key object and then use them as that object.

Please consider breaking this into multiple tickets so that we can merge the simple part and don't delay them until this whole ticket is merge.

Please check my comments and resubmit.

Full coverage using automated tests is very important so that in future version we don't introduce regressions.

The development team is small and volatile so automated tests are the only way to preserver the knowledge about how the code should behave.

Many thanks for your contribution!

**Comment by twisted-trac** [comment]
|[<img alt="the0id's avatar" src="https://avatars.githubusercontent.com/u/19194467?s=50" width="50" height="50">](https://github.com/the0id)<a name="note_19"></a>|@the0id commented|
|-|-|

Originally I agreed about doing this in sections because ECDSA wasn't working at the time, but now that it is, I'm thinking it might be better to do ECDSA with ECDH.  This is because I've been having trouble maintaining 2 (3 if you count curve25519) separate branches, and trying to develop them while trying to keep them inline with each other and the main trunk.

This isn't a complaint, it's how things go, but it has caused a headache that I think we can avoid if we treat ECDH and ECDSA as one PR. Though since there are two tickets open and they don't completely correspond to the branch(es) on github, I'm not sure what's the best way to proceed.

Which makes it a little tricky to answer your questions because it depends on which branch of the code you're talking about, but I think many of them can be answered in the github conversation for the PR: https://github.com/twisted/twisted/pull/432

I know it makes things even more confusing (and more of an argument to close this ticket and 8730 and create a new, single ticket), but I would completely ignore the changes in this TRAC ticket because they are completely obsolete now.

To help here are how the 3 branches of I have in github correspond to the EC* changes.

https://github.com/the0id/twisted/tree/8370-the0id-sshECH-redux -> JUST ECDH, no ECDSA.  Most minimal change set, and what is currently part of PR 432.

https://github.com/the0id/twisted/tree/8370-the0id-sshECDSA -> ECDH + ECDSA.  Until a few days ago when I decided to check back in on this project, I had a Eureka moment and fixed what was wrong with ECDSA.  ___This is the branch I would now recommend we use to bring EC* to Twisted_.__

https://github.com/the0id/twisted/tree/8370-the0id-sshEC-plus-Curve25519 -> ECDH + ECDSA + Curve25519.  This is what I did because I was on a roll and because OpenSSH now defaults to Curve25519.  The issue with this one is that Cryptography does not support Curve25519 at this time, so I had to find two other external libraries (one for ECDH and one to sign) to implement Curve25519.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_20"></a>|@glyph commented|
|-|-|

It looks like acabhishek942 has taken a stab at responding to review feedback; this work is continued in https://github.com/twisted/twisted/pull/539

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_21"></a>|@glyph commented|
|-|-|

[#8811](https://github.com/twisted/twisted/issues/8811) was closed as a duplicate of this.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_22"></a>|@glyph commented|
|-|-|

> Originally I agreed about doing this in sections because ECDSA wasn't working at the time, but now that it is, I'm thinking it might be better to do ECDSA with ECDH. This is because I've been having trouble maintaining 2 (3 if you count curve25519) separate branches, and trying to develop them while trying to keep them inline with each other and the main trunk.

As Cryptography puts it [in their contributing documentation](https://cryptography.io/en/latest/development/submitting-patches/):

> > “Patches should be small to facilitate easier review. [Studies have shown that review quality falls off as patch size grows.](https://smartbear.com/learn/code-review/best-practices-for-peer-code-review/) Sometimes this will result in many small PRs to land a single large feature.”

_Just_ the additions to `Key` in this case are larger than 400 lines, the maximum recommended reviewable size.  Smaller changes get reviewed faster, and are easier to spot problems in.  We aren't as strict about change size on Twisted as Cryptography is, but given that this _is_ crypto code, we have cause to be just as concerned as they are.  So keeping this in manageably sized chunks is really a requirement.  Luckily acabhishek942 seems to be taking this work on, so hopefully we'll have something to merge :).

**Comment by twisted-trac** [comment]
|[<img alt="the0id's avatar" src="https://avatars.githubusercontent.com/u/19194467?s=50" width="50" height="50">](https://github.com/the0id)<a name="note_23"></a>|@the0id commented|
|-|-|

I saw that acabhishek942's changes have been merged into the trunk so I went ahead and tried to merge his changes with mine.

For the most part I liked the changes, but I did run into a few issues when trying to use ECDSA.  I know neither of his PRs is intended to support it yet, but it looked like some of the changes that were submitted will either cause problems in the future, or I (likely) did something wrong while merging :).

This is why I put the review flag back on this ticket.  I don't expect this PR to be merged, but I've updated my PR with a combination of my and acabhishek942's changes that seems to work for ECDH and ECDSA.  It's testing right now, and it's likely to fail if not for any other reason than covecov has not been pleased with any of my offerings, but I offer it up in case the team has the same issues I had.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_24"></a>|@glyph removed owner|
|-|-|

Hey the0id,

Thanks again for your contributions here.  acabhishek942's changes were derived from yours, so we should be able to re-integrate ECDSA support; the problems shouldn't be fundamental.

A point of order, though: If you don't expect the PR to be merged, then it's not ready for review.  Review is a specific workflow state.  If you just want some discussion, please mention it on IRC, or on the mailing list.  The reason for this discrete workflow state is to try to focus project member energy as much as possible on changes that should be ready to be integrated.  This is hard enough without not-sure-what-state-it's-really-in tickets like this one; I've tripped over this one a few times in the last few days and pondered what to do about it :-).

As such, I'm going to remove the review keyword for now.  If you, or anyone else, gets it back into a mergeable state, please resubmit it for review.  I'm leaving it open, since if another ticket completes the full pile of ECDH work, then this should be closed even if it's a PR for some smaller unit of functionality that closes the loop, since the overall _goal_ of this ticket may be broken down and satisfied by a series of smaller changes.

I'm also un-assigning it because if you don't have time to complete it there's no particular reason it should remain assigned to you.

**Comment by twisted-trac** [comment]
|[<img alt="the0id's avatar" src="https://avatars.githubusercontent.com/u/19194467?s=50" width="50" height="50">](https://github.com/the0id)<a name="note_25"></a>|@the0id commented|
|-|-|

Not a problem.  The reason why I brought it up and kept it in review was because I made some changes that might be relevant and wanted to bring it to your attention.

I know things have become muddled with my multiple branches and tickets for this so I just wanted to clear it up.

[[| PR 432 ](https://github.com/twisted/twisted/pull/432)] Is the consolidation of the ECDH and ECDSA efforts, which makes ticket [[[[#8370](https://github.com/twisted/twisted/issues/8370)](https://github.com/twisted/twisted/issues/8370) | #8370 ]] obsolete, despite what it says above on this ticket, and I'll close it.

The [[| Curve25519 ](https://github.com/the0id/twisted/tree/8370-the0id-sshEC-plus-Curve25519)] branch is still being updated as needed, though as I mentioned before I'm keeping it separate and not making a PR because of the 3rd party libraries it requires.

PR 432 passes all but the code coverage checks, but as my mantra goes, I probably won't finish the tests, so do with it as you like. :)

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_26"></a>|@adiroiban commented|
|-|-|

...

**Comment by twisted-trac** [comment]
|[<img alt="acabhishek942's avatar" src="https://avatars.githubusercontent.com/u/7778186?s=50" width="50" height="50">](https://github.com/acabhishek942)<a name="note_27"></a>|@acabhishek942 commented|
|-|-|

...

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_28"></a>|@glyph set owner to @acabhishek942|
|-|-|

I'm a little confused about the state here.  It looks like I took this out of review, the0oid said they weren't going to do the work, and then acabhishek942 put it back into review, but there are no commits on PR 432 from acabhishek942.

Just from glancing at that PR, I can see that there is still a coverage issue that needs to be addressed, so please fix that before resubmitting (if that is the PR that is supposed to be in review here).

Also, that PR includes lots of random cleanup changes (re-indenting, etc).  I'd love it if we could split out a "random cleanups" branch which could probably be merged with just a skim, and thereby make the diff for the more substantive crypto-stuff PR shorter and easier to review by itself.

But generally please just make it clear what ought to be reviewed here with a comment, and resubmit for review :).

**Comment by twisted-trac** [comment]
|[<img alt="the0id's avatar" src="https://avatars.githubusercontent.com/u/19194467?s=50" width="50" height="50">](https://github.com/the0id)<a name="note_29"></a>|@the0id set owner to @the0id|
|-|-|

I open a ticket and say it's not finished.\\
I open another ticket saying it's a little more finished than the first ticket.\\
I finish the second ticket, commit the changes to the first ticket.\\
You remove the review flag, acabhishek942 sets it.\\
Markwilliams merges the trunk to the PR.\\
I say I won't have time to work on it any more, I find time.\\
I say I won't write unit tests, I write unit tests.\\

Everything you believed to be true is a lie.
What's confusing about that?

If txchecker and I become friends, we might even have full code coverage.

To sum it up, I had some extra time today and seeing how there seems to be activity on this branch/PR, I decided to give unit tests another shot.  And maybe it's because I'm smarter today than I was yesterday, they seemed fairly easy to do (I say before the checks are completed).

I'm not too sure I'm following your "random cleanups" thought completely.  I get the general idea that, yes, there are lots of small changes as I've muddled my way through this, but I'd worry that doing so would take more time than there'd be benefit, and has the chance of introducing new errors.

As I type this the tests are currently running and so I'm going to take a break.  I won't put the review flag back up just yet but since my old mantra of, "I don't have time for unit tests" is no longer true, I'll fall back to my first one of, "Do with this as you will." :)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_30"></a>|@glyph commented|
|-|-|

> Everything you believed to be true is a lie. What's confusing about that?

Clear as mud :).

> I'm not too sure I'm following your "random cleanups" thought completely. I get the general idea that, yes, there are lots of small changes as I've muddled my way through this, but I'd worry that doing so would take more time than there'd be benefit, and has the chance of introducing new errors.

Research shows that change size is correlated with defect count, so I am always trying to get changes to be as small as humanly possible.  [Cryptography has gone so far as to make this an official part of their review process](https://cryptography.io/en/latest/development/submitting-patches/?highlight=contributing#submitting-patches), and I'm thinking of doing the same.

If you find the Git interface to do this confusing I could volunteer to pull out the cleanups into a separate PR myself once it's ready for review again.

**Comment by twisted-trac** [comment]
|[<img alt="the0id's avatar" src="https://avatars.githubusercontent.com/u/19194467?s=50" width="50" height="50">](https://github.com/the0id)<a name="note_31"></a>|@the0id commented|
|-|-|

I see why you want the changes to be as small as possible, and you're right, but in this case I have a feeling that trying to break this up into smaller changes will increase the chances of errors.

All of the changes are connected to each other, so to push small changes one at a time, I believe the code will have to be changed to compensate so it won't throw an error when other code it relies on isn't present, which then will have to be changed back once the rest of the code is merged.  Plus with txchecker enforcing strict code styling, I don't think it will be possible to remove the random cleanups and have it pass the checks.

Since the unthinkable has happened and this PR meets all of the requirements for a review, I'm putting the keyword back.  If it still looks like we need to break this up after the review then I can look into how to go about that.

Also I believe some of the code in my PR was actually written by acabhishek942, and credit should go where it's due.  Is there a way to give acabhishek942 credit for those sections?

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_32"></a>|@rodrigc set status to `closed`|
|-|-|

In changeset e116bbe34dc3a3855a674043bc5f30d7e6417c7f
```
#!CommitTicketReference repository="" revision="e116bbe34dc3a3855a674043bc5f30d7e6417c7f"
Merge pull request #630 from twisted/8931-rodrigc-ssh

Author: the0id , acabhishek942
Reviewer: rodrigc
Fixes: ticket:8325 , ticket:8784 , ticket:8811 ,   ticket:8894 , ticket:8931   

Add support for elliptic curve (EC) keys to conch
```

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_33"></a>|@rodrigc commented|
|-|-|

https://github.com/twisted/twisted/commit/e116bbe34dc3a3855a674043bc5f30d7e6417c7f

**Comment by twisted-trac** [comment]
|[<img alt="Julian's avatar" src="https://avatars.githubusercontent.com/u/329822?s=50" width="50" height="50">](https://github.com/Julian)<a name="note_34"></a>|@Julian commented|
|-|-|

Is there a ticket to track adding support for ed25519?

**Comment by twisted-trac** [comment]
|[<img alt="the0id's avatar" src="https://avatars.githubusercontent.com/u/19194467?s=50" width="50" height="50">](https://github.com/the0id)<a name="note_35"></a>|@the0id commented|
|-|-|

Replying to [Julian Berman](#note_34):
> Is there a ticket to track adding support for ed25519?

Yes [[[#8966](https://github.com/twisted/twisted/issues/8966)](https://github.com/twisted/twisted/issues/8966) #8966].

And there's a PR [here](https://github.com/twisted/twisted/pull/644).
